src/integer_mod_q/mat_zq/trace.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright © 2023 Marcel Luca Schmidt
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module contains the implementation of the `trace` function.
+
+use flint_sys::fmpz_mod_mat::fmpz_mod_mat_trace;
+
+use super::MatZq;
+use crate::{
+    error::MathError,
+    integer_mod_q::Zq,
+    traits::{GetNumColumns, GetNumRows},
+};
+
+impl MatZq {
+    /// Returns the trace of a matrix and an error,
+    /// if the matrix is not square.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let matrix = MatZq::from_str("[[1,2],[3,4]] mod 5").unwrap();
+    /// let trace = matrix.trace().unwrap();
+    /// ```
+    pub fn trace(&self) -> Result<Zq, MathError> {
+        // check if matrix is square
+        if self.get_num_rows() != self.get_num_columns() {
+            return Err(MathError::NoSquareMatrix(self.to_string()));
+        }
+
+        let mut out = Zq::try_from_int_int(0, self.get_mod()).unwrap();
+        unsafe {
+            fmpz_mod_mat_trace(&mut out.value.value, &self.matrix);
+        }
+        Ok(out)
+    }
+}
+
+#[cfg(test)]
+mod test_trace {
+
+    use crate::integer_mod_q::{MatZq, Zq};
+    use std::str::FromStr;
+
+    /// Test whether `trace` correctly calculates the trace of a matrix
+    #[test]
+    fn trace_works() {
+        let mat1 = MatZq::from_str("[[5,2,0],[2,1,0],[0,0,1]] mod 10").unwrap();
+        let mat2 = MatZq::from_str("[[-1,0],[0,1]] mod 2").unwrap();
+
+        let trace1 = mat1.trace().unwrap();
+        let trace2 = mat2.trace().unwrap();
+
+        assert_eq!(Zq::try_from_int_int(7, 10).unwrap(), trace1);
+        assert_eq!(Zq::try_from_int_int(0, 2).unwrap(), trace2);
+    }
```

Comment:
add a test that oversteps the modulus
```suggestion
        let mat1 = MatZq::from_str("[[5,2,0],[2,8,0],[0,0,4]] mod 10").unwrap();
        let mat2 = MatZq::from_str("[[-1,0],[0,1]] mod 2").unwrap();

        let trace1 = mat1.trace().unwrap();
        let trace2 = mat2.trace().unwrap();

        assert_eq!(Zq::try_from_int_int(7, 10).unwrap(), trace1);
        assert_eq!(Zq::try_from_int_int(0, 2).unwrap(), trace2);
    }
```